staticcheck/lint.go

```diff
@@ -2788,3 +2789,87 @@ func (c *Checker) CheckMissingEnumTypesInDeclaration(j *lint.Job) {
 		ast.Inspect(f, fn)
 	}
 }
+
+func (c *Checker) CheckUnreachableTypeCases(j *lint.Job) {
+	// Check if T subsumes V in a type switch:
+	// T and V are structurally identical interfaces;
+	// interface T is a superset of V;
+	// T is an interface implemented by concrete type V.
+	subsumes := func(T, V types.Type) bool {
+		tIface, ok := T.Underlying().(*types.Interface)
+		if !ok {
+			return false
+		}
+
+		return types.Implements(V, tIface)
+	}
+
+	intersect := func(Ts, Vs []types.Type) (types.Type, types.Type, bool) {
```

Comment:
Let's not call this `intersect`; it's not really an intersection. Maybe `subsumesAny`?